compose/service.py

```diff
@@ -590,17 +590,51 @@ def build_volume_binding(volume_spec):
     return os.path.abspath(os.path.expandvars(external)), internal
 
 
+def add_port_mapping(port_bindings, internal_port, external):
+    if internal_port in port_bindings:
+        port_bindings[internal_port].append(external)
+    else:
+        port_bindings[internal_port] = [external]
+
+
 def build_port_bindings(ports):
     port_bindings = {}
     for port in ports:
-        internal_port, external = split_port(port)
-        if internal_port in port_bindings:
-            port_bindings[internal_port].append(external)
+        internal_port_range, external_range = split_port(port)
+        if external_range is None:
+            for internal_port in internal_port_range:
+                add_port_mapping(port_bindings, internal_port, None)
         else:
-            port_bindings[internal_port] = [external]
+            for internal_port, external_port in zip(internal_port_range, external_range):
+                add_port_mapping(port_bindings, internal_port, external_port)
     return port_bindings
 
 
+def to_port_range(port):
+    if not port:
+        return None
+
+    protocol = ""
+    if "/" in port:
+        parts = port.split("/")
+        if len(parts) != 2:
+            raise ConfigError('Invalid port "%s", should be '
+                              '[[remote_ip:]remote_port:]port[/protocol]' % port)
+        port, protocol = parts
+        protocol = "/" + protocol
+
+    parts = str(port).split('-')
+
+    if len(parts) == 1:
+        return ["%s%s" % (port, protocol)]
+
+    if len(parts) == 2:
+        return ["%s%s" % (p, protocol) for p in range(int(parts[0]), int(parts[1]) + 1)]
```

Comment:
Looks like Docker does indeed do it client-side, so that's where we'll need to do it too. Moving it to the server is a discussion to be had on docker/docker.

Paging @shin-: does processing of port ranges belong in docker-py, in your opinion?